### PR TITLE
Restart Based Search and Policy Gradient Method 

### PR DESCRIPTION
**Pull request by 3rdCore**
This is my proposition for the implementation of the _Restart Based Search Strategy_. It implements 3 different restart based strategies : `staticRBS` , `geometricRBS` and `LubyRBS`, that will define how the stopping criteria value `L` will increase as the research progresses, the user define the initial value  `L0` : 

1. `staticRBS` : the value `L` remains equal to `L0` for each restart.  
2. `geometricRBS` : the value `L` is increased by the geometric factor `α` at each restart. 
3. `LubyRBS` : the value `L` is multiplied by the factor `Luby[i]` for the i-th restart. The Luby sequence is a sequence of the following form: 1,1,2,1,1,2,4,1,1,2,1,1,2,4,8, . .and gives theoretical improvement on the search in the general case.

I also proposed 3 different stopping criteria based on several statistics of the search : 

1.  `InfeasibleNodeCriteria` : the search stops and restarts as soon as we reached `L` Infeasible state.
2.  `SolutionFoundCriteria`: the search stops and restarts as soon as we reached `L` solution of the CP problem. 
3.  `VisitedNodeCriteria`: the search stops and restarts as soon as we visited L nodes ( stops at the node L+1 ). 

RBS strategy should be used with stochastic policy otherwise, at each restart, the agent will follow the same path as before. Hence, one can find examples using a Policy Gradient based agent in the testset of `rbs.jl`. A PR on [SeaPearlZoo](https://github.com/corail-research/SeaPearlZoo) will come next to provide clear example using PPO. 

**What could be done next ?** 
- [ ] implement the _Failure Based Search Strategy_

### CODE REVIEW

File: src/CP/core/model.jl [hunk 1]
```diff
@@ -1,17 +1,22 @@
...
-    numberOfSolutions       ::Int
-    solutions               ::Vector{Solution}
-    nodevisitedpersolution  ::Vector{Int}
-    objectives              ::Union{Nothing, Vector{Int}}
+    numberOfNodes                           ::Int
+    numberOfSolutions                       ::Int
+    numberOfInfeasibleSolutions             ::Int
+    numberOfSolutionsBeforeRestart          ::Int
+    numberOfInfeasibleSolutionsBeforeRestart::Int
+    numberOfNodesBeforeRestart              ::Int
```
Comment (3rdCore) [inline src/CP/core/model.jl:11]: Usefull statistics to define the stopping criteria

File: src/CP/core/search/ilds.jl [hunk 2]
```diff
@@ -2,19 +2,19 @@
...
-function initroot!(toCall::Stack{Function}, ::Type{ILDSearch}, model::CPModel, variableHeuristic::AbstractVariableSelection, valueSelection::ValueSelection)
+function initroot!(toCall::Stack{Function}, ::ILDSearch , model::CPModel, variableHeuristic::AbstractVariableSelection, valueSelection::ValueSelection)
     isboundedlist = [!isbound(v) for (k,v) in model.variables]
     @assert !isempty(isboundedlist) "initialisation failed : no declared variables"
     depth = sum(isboundedlist)
     
     # Note that toCall stack is a LIFO data structure, expandIlds with a discrepancy threshold of 0 will be the first one to execute (then with 1, 2, 3, etc.)
     for k in depth:-1:1
-        push!(toCall, (model) -> (expandIlds!(toCall,k,depth, nothing, model, variableHeuristic, valueSelection)))
+        push!(toCall, (model) -> (restart_search!(model); expandIlds!(toCall,k,depth, nothing, model, variableHeuristic, valueSelection)))
```
Comment (3rdCore) [inline src/CP/core/search/ilds.jl:17]: The call to `restart_search!` reset some metrics and ensure that we are back at the 
```suggestion
        push!(toCall, (model) -> (restart_search!(model); expandIlds!(toCall,k,depth, nothing, model, variableHeuristic, valueSelection)))top of the tree.
```

File: src/CP/core/search/rbs.jl [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+function that expand the search tree. In restart based strategy, we fill the Stack with calls to expandRbs with different nodeLimit. The nodeLimit corresponds 
+to the number of infeasiblesolution that can be reached before restarting the search a the top of the tree with a possibly different nodeLimit. This search 
+strategy requires the use of a stochastic variable/value heuristic, otherwise, at each restart the search will end-up on the exact previous solutions. 
+
+geometricRBSearch : At each restart, the number of infeasible solution before restart is increased by the geometric factor strategy.α. strategy.L states the 
+initial number of infeasible solution for the first search.  
+"""
+function initroot!(toCall::Stack{Function}, strategy::geometricRBSearch, model::CPModel, variableHeuristic::AbstractVariableSelection, valueSelection::ValueSelection)
+    nodeLimit = strategy.L
+    for i in strategy.n:-1:2
```
Comment (3rdCore) [inline src/CP/core/search/rbs.jl:32]: The for loop is reverse as we works with a Stack

File: src/CP/core/search/rbs.jl [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+function Luby(n::Int64)
+    output=zeros(n)
+    K=map(x -> 2^x-1,1:n)   
+    lowerbound= K[(map(x->findfirst(K->K>x,K),1:n).-1)]
+    for x in 1:n 
+    output[x]= (x in K) ? (x+1)/2 : output[x-lowerbound[x]]
+    end
+    return Int.(output)
+end
```
Comment (3rdCore) [inline src/CP/core/search/rbs.jl:68]: This works fine, don't bother you to understand this blackboxy code.

File: src/CP/core/search/strategies.jl [hunk 5]
```diff
@@ -1,4 +1,77 @@
...
+be specified in the search! function. The search strategy choose how the _toCall_ Stack will be filled and empty ( ie. which branch will 
+be explored at a certain node ).
+"""    
 abstract type SearchStrategy end
 
-abstract type DFSearch <: SearchStrategy end
-abstract type ILDSearch <: SearchStrategy end
+struct DFSearch <: SearchStrategy end
+
+struct ILDSearch <: SearchStrategy end
```
Comment (3rdCore) [inline src/CP/core/search/strategies.jl:9]: I switched back from type to struct as we needed to declare arguments for the different  ___RBSearch such as `L`.

File: src/CP/valueselection/learning/environment.jl [hunk 6]
```diff
@@ -41,3 +43,38 @@ function (learner::DQNLearner)(env::AbstractCPEnv{TS}) where {TS <: NonTabularTr
...
+        send_to_host 
+    output[findall(mask)] = softmax(logits[findall(mask)])
+    if p.update_step < p.n_random_start
+        
+            Categorical(fill(1 / length(x), length(x)); check_args = false)
+        
+    else
+        Categorical(x; check_args = false)
+    end
+end
```
Comment (3rdCore) [inline src/CP/valueselection/learning/environment.jl:80]: This is no longer usefull.

File: src/CP/valueselection/learning/environment.jl [hunk 7]
```diff
@@ -41,3 +43,38 @@ function (learner::DQNLearner)(env::AbstractCPEnv{TS}) where {TS <: NonTabularTr
...
+
+function (learner::A2CLearner)(env::CPEnv{ST}) where {ST <: NonTabularTrajectoryState}
+    env |>
+    state |>
+    x ->
+        send_to_device(device(learner), x) |>
+        learner.approximator.actor |>
+        vec |>
+        send_to_host
+end
```
Comment (3rdCore) [inline src/CP/valueselection/learning/environment.jl:55]: We needed to redefine the forward pass, to work with `NonTabularTrajectoryState`

File: src/CP/core/search/rbs.jl [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    # Dealing with limits
+    model.statistics.numberOfNodes += 1
+    model.statistics.numberOfNodesBeforeRestart += 1
+
+    if !belowNodeLimit(model)
+        return :NodeLimitStop
+    end
+    if !belowSolutionLimit(model)
+        return :SolutionLimitStop
+    end
```
Comment (PierreTsr) [inline src/CP/core/search/rbs.jl:91]: We still use these limits with RBS?

File: src/CP/valueselection/learning/environment.jl [hunk 9]
```diff
@@ -41,3 +43,38 @@ function (learner::DQNLearner)(env::AbstractCPEnv{TS}) where {TS <: NonTabularTr
         vec |>
         send_to_host
 end
+
+function (learner::A2CLearner)(env::CPEnv{ST}) where {ST <: NonTabularTrajectoryState}
```
Comment (PierreTsr) [inline src/CP/valueselection/learning/environment.jl:47]: `CPEnv` => `AbstractCPEnv` ?

File: src/CP/valueselection/learning/environment.jl [hunk 10]
```diff
@@ -41,3 +43,38 @@ function (learner::DQNLearner)(env::AbstractCPEnv{TS}) where {TS <: NonTabularTr
...
+    x ->
+        send_to_device(device(learner), x) |>
+        learner.approximator.actor |>
+        vec |>
+        send_to_host
+end
+
+function Array(state::T) where {T<:NonTabularTrajectoryState}
+    state
+end
```
Comment (PierreTsr) [inline src/CP/valueselection/learning/environment.jl:59]: What is the use of this function?

File: src/CP/valueselection/learning/environment.jl [hunk 11]
```diff
@@ -41,3 +43,38 @@ function (learner::DQNLearner)(env::AbstractCPEnv{TS}) where {TS <: NonTabularTr
...
+
+function RLBase.prob(p::PPOPolicy, env::CPEnv{ST}) where {ST <: NonTabularTrajectoryState}
+    s = state(env)
+    mask = isnothing(legal_action_space_mask(env)) ? nothing : legal_action_space_mask(env)
+    prob(p, s, mask)
```
Comment (PierreTsr) [inline src/CP/valueselection/learning/environment.jl:64]: this could be simplified:
`mask = legal_action_space_mask(env)`

File: src/CP/valueselection/learning/environment.jl (hunk 6 above)
Comment (PierreTsr) [reply src/CP/valueselection/learning/environment.jl:80]: Will it be removed then?

File: src/RL/nn_structures/nn_structures.jl [hunk 12]
```diff
@@ -1,4 +1,6 @@
-using GraphSignals
+using Flux
+using Zygote
+using CUDA
```
Comment (PierreTsr) [inline src/RL/nn_structures/nn_structures.jl:3]: Aren't they already included before?

File: src/experiment/benchmark_solving.jl [hunk 13]
```diff
@@ -19,13 +19,13 @@ This function might evolve for a more precise one which will provide better anal
 function benchmark_solving(;
         valueSelectionArray::Union{T, Array{T, 1}}, 
         generator::AbstractModelGenerator,
-        strategy::Type{DFSearch}=DFSearch,
+        strategy::S=DFSearch,
```
Comment (PierreTsr) [inline src/experiment/benchmark_solving.jl:22]: Shouldn't it be `DFSearch()`?

File: test/CP/core/search/rbs.jl [hunk 14]
```diff
@@ -0,0 +1,567 @@
...
+            Dict("x1" => 2, "x2" => 1, "x3" => 2, "x4" => 4)
+        ]
+
+        for solution in model.statistics.solutions
+            @test solution in possible_solutions
+        end
+
+    end
+
+end
```
Comment (PierreTsr) [inline test/CP/core/search/rbs.jl:567]: I trust you on this ^^

**Review by PierreTsr — CHANGES_REQUESTED** [review]
Very impressive work. I'm eager to see how it will perform! We should discuss a few points before mergring though ;)

File: src/RL/nn_structures/nn_structures.jl (hunk 12 above)
Comment (3rdCore) [reply src/RL/nn_structures/nn_structures.jl:3]: well, I sincerely think that there is some kind of paranormal activity going on dependencies hahah

**Review by qcappart — APPROVED** [review]
Great work ! I am very eager to see the results on our different RL-based methods :-)

I approve the PR, but you can resolve the comment of @PierreTsr before the merging.

**Review by PierreTsr — APPROVED** [review]
Awesome, I'm good with all the new changes ! (just the `Array` function is still here, is it normal?)

File: src/CP/core/search/search.jl [hunk 15]
```diff
@@ -38,8 +42,8 @@ function search!(model::CPModel, ::Type{Strategy}, variableHeuristic::AbstractVa
         currentProcedure = pop!(toCall)
         currentStatus = currentProcedure(model)
     end
-
     # set final reward and last observation
+    removeSolutionDoublons(model)
```
[the remark this answers is not in the document]
Comment (3rdCore) [reply src/CP/core/search/search.jl:46]: yes, simply because during a DFSearch, no solution was encountered twice.

**Comment by 3rdCore** [comment]
> 
> 
> Awesome, I'm good with all the new changes ! (just the `Array` function is still here, is it normal?)

Yes, it is needed by the update! function of the A2C learner !
